datafusion/core/tests/sql/parquet_schema.rs

```diff
@@ -99,6 +99,100 @@ async fn schema_merge_ignores_metadata_by_default() {
     assert_no_metadata(&actual);
 }
 
+#[tokio::test]
+async fn merge_compatible_schemas() -> Result<()> {
+    let tmp_dir = TempDir::new()?;
+
+    let schema1 = Schema::new_with_metadata(
+        vec![
+            Field::new("a", DataType::Int8, false),
+            Field::new("b", DataType::Int16, false),
+            Field::new("c", DataType::Int32, false),
+        ],
+        HashMap::new(),
+    );
+
+    let schema2 = Schema::new_with_metadata(
+        vec![
+            Field::new("a", DataType::Int64, false),
+            Field::new("b", DataType::Int32, false),
+            Field::new("c", DataType::Int16, false),
+        ],
+        HashMap::new(),
+    );
+
+    create_parquet_file(tmp_dir.path(), "t1.parquet", &schema1).await?;
+    create_parquet_file(tmp_dir.path(), "t2.parquet", &schema2).await?;
+
+    let ctx = SessionContext::default();
+
+    // with type coercion disabled
+    let err = ctx
+        .register_parquet(
+            "test",
+            tmp_dir.path().to_str().unwrap(),
+            ParquetReadOptions::default().with_coerce_types(false),
+        )
+        .await
+        .expect_err("should fail");
+    assert!(err
+        .to_string()
+        .contains("Schema merge failed due to different, but compatible, data types"));
+
+    // with type coercion enabled
+    ctx.register_parquet(
+        "test",
+        tmp_dir.path().to_str().unwrap(),
+        ParquetReadOptions::default().with_coerce_types(true),
+    )
+    .await?;
+
+    let df = ctx.table("test")?;
+
+    let expected_schema = Schema::new_with_metadata(
+        vec![
+            Field::new("a", DataType::Int64, false),
+            Field::new("b", DataType::Int32, false),
+            Field::new("c", DataType::Int32, false),
+        ],
+        HashMap::new(),
+    );
+
+    let actual_schema: Schema = df.schema().into();
+    assert_eq!(expected_schema, actual_schema);
+
+    let batches = df.collect().await?;
```

Comment:
there is also the `arrow_typeof` function added by @waitingkuo , so you could verify the expected schema with a query like

```sql
select arrow_typeof(a), arrow_typeof(b), arrow_typeof(c) from test;
```